freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/UmsUsersStateProvider.java

```diff
@@ -54,92 +77,232 @@
+
+    public Map<String, UmsUsersState> getEnvToUmsUsersStateMap(
+            String accountId, String actorCrn, Collection<String> environmentCrns,
+            Set<String> userCrns, Set<String> machineUserCrns, Optional<String> requestIdOptional) {
         try {
             LOGGER.debug("Getting UMS state for environments {} with requestId {}", environmentCrns, requestIdOptional);
 
-            Map<String, UmsUsersState> envUsersStateMap = new HashMap<>();
-
             boolean fullSync = userCrns.isEmpty() && machineUserCrns.isEmpty();
 
-            List<User> users = getUsers(actorCrn, accountId, requestIdOptional, fullSync, userCrns);
-
-            List<MachineUser> machineUsers = getMachineUsers(actorCrn, accountId, requestIdOptional, fullSync, machineUserCrns);
-
-            Map<String, FmsGroup> crnToFmsGroup = grpcUmsClient.listGroups(actorCrn, accountId, List.of(), requestIdOptional).stream()
-                    .collect(Collectors.toMap(Group::getCrn, this::umsGroupToGroup));
-
-            Map<WorkloadAdministrationGroup, FmsGroup> wags = grpcUmsClient.listWorkloadAdministrationGroups(INTERNAL_ACTOR_CRN, accountId, requestIdOptional)
-                    .stream()
-                    .collect(Collectors.toMap(wag -> wag, wag -> nameToGroup(wag.getWorkloadAdministrationGroupName())));
-            environmentCrns.forEach(environmentCrn -> {
-                UmsUsersState.Builder umsUsersStateBuilder = new UmsUsersState.Builder()
-                        .setWorkloadAdministrationGroups(wags.values());
-
-                UsersState.Builder usersStateBuilder = new UsersState.Builder();
-                Set<String> wagNamesForOtherEnvironments = new HashSet<>();
-
-                crnToFmsGroup.values().forEach(usersStateBuilder::addGroup);
-
-                // Only add workload admin groups that belong to this environment.
-                // At the same time, build a set of workload admin groups that are
-                // associated with other environments so we can filter these out in
-                // the per-user group listing in handleUser.
-                wags.entrySet().forEach(wagEntry -> {
-                    WorkloadAdministrationGroup wag = wagEntry.getKey();
-                    String groupName = wag.getWorkloadAdministrationGroupName();
-                    if (wag.getResource().equalsIgnoreCase(environmentCrn)) {
-                        usersStateBuilder.addGroup(wagEntry.getValue());
-                    } else {
-                        Crn resourceCrn = getCrn(wag);
-                        if (resourceCrn != null && resourceCrn.getService() == Crn.Service.ENVIRONMENTS
-                                && resourceCrn.getResourceType() == ResourceType.ENVIRONMENT) {
-                            wagNamesForOtherEnvironments.add(groupName);
-                        }
-                    }
-                });
+            Map<String, UmsUsersState.Builder> envUsersStateMap;
```

Comment:
nit: Any reason `getUmsUsersStateMapBulk` and `getUmsUsersStateMap` needs to
return a `Builder` instead of the built `UmsUserState`?

Reply:
nope. fixed